// ngx-datetime-range-picker/esm2015/lib/ngx-datetime-range-picker.module.js
import { NgModule, Optional, SkipSelf } from "@angular/core";
import { CommonModule } from "@angular/common";
import { FormsModule } from "@angular/forms";
import { MaterialModule } from "./material/material.module";
import { NgxDatetimeRangePickerComponent } from "./ngx-datetime-range-picker.component";
import { ObjNgFor } from "./pipes/objNgFor.pipe";
import { NgxDatetimeRangePickerService } from "./ngx-datetime-range-picker.service";
import * as ɵngcc0 from '@angular/core';
export class NgxDatetimeRangePickerModule {
    constructor(parentModule) {
        if (parentModule) {
            throw new Error(`ERR_NGX_DATETIME_RANGE_PICKER:
        NgxDatetimeRangePickerModule is already loaded. Import it in the AppModule only`);
        }
    }
    static forRoot() {
        return {
            ngModule: NgxDatetimeRangePickerModule,
            providers: [NgxDatetimeRangePickerService]
        };
    }
}
NgxDatetimeRangePickerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxDatetimeRangePickerModule });
NgxDatetimeRangePickerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxDatetimeRangePickerModule_Factory(t) { return new (t || NgxDatetimeRangePickerModule)(ɵngcc0.ɵɵinject(NgxDatetimeRangePickerModule, 12)); }, imports: [[CommonModule, FormsModule, MaterialModule], MaterialModule] });
NgxDatetimeRangePickerModule.ctorParameters = () => [
    { type: NgxDatetimeRangePickerModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxDatetimeRangePickerModule, { declarations: function () { return [ObjNgFor, NgxDatetimeRangePickerComponent]; }, imports: function () { return [CommonModule, FormsModule, MaterialModule]; }, exports: function () { return [NgxDatetimeRangePickerComponent, MaterialModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxDatetimeRangePickerModule, [{
        type: NgModule,
        args: [{
                declarations: [ObjNgFor, NgxDatetimeRangePickerComponent],
                imports: [CommonModule, FormsModule, MaterialModule],
                exports: [NgxDatetimeRangePickerComponent, MaterialModule]
            }]
    }], function () { return [{ type: NgxDatetimeRangePickerModule, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }]; }, null); })();
