// ngx-datetime-range-picker/esm2015/lib/material/material.module.js
import { NgModule } from "@angular/core";
import { MatButtonModule } from "@angular/material/button";
import { MatFormFieldModule } from "@angular/material/form-field";
import { MatIconModule } from "@angular/material/icon";
import { MatInputModule } from "@angular/material/input";
import { MatSelectModule } from "@angular/material/select";
import * as ɵngcc0 from '@angular/core';
const modules = [MatFormFieldModule, MatButtonModule, MatInputModule, MatIconModule, MatSelectModule];
export class MaterialModule {
}
MaterialModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MaterialModule });
MaterialModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MaterialModule_Factory(t) { return new (t || MaterialModule)(); }, imports: [modules, MatFormFieldModule, MatButtonModule, MatInputModule, MatIconModule, MatSelectModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MaterialModule, { imports: [MatFormFieldModule, MatButtonModule, MatInputModule, MatIconModule, MatSelectModule], exports: [MatFormFieldModule, MatButtonModule, MatInputModule, MatIconModule, MatSelectModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MaterialModule, [{
        type: NgModule,
        args: [{
                imports: modules,
                exports: modules
            }]
    }], null, null); })();
