// ngx-mat-datetime-picker/esm5/lib/core/ngx-mat-native-date.module.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __decorate } from "tslib";
import { PlatformModule } from '@angular/cdk/platform';
import { NgModule } from '@angular/core';
import { MAT_DATE_FORMATS } from '@angular/material/core';
import { NgxMatDateAdapter } from './ngx-mat-date-adapter';
import { NgxMatNativeDateAdapter } from './ngx-mat-native-date-adapter';
import { NGX_MAT_NATIVE_DATE_FORMATS } from './ngx-mat-native-date-formats';
import * as ɵngcc0 from '@angular/core';
var NgxNativeDateModule = /** @class */ (function () {
    function NgxNativeDateModule() {
    }
NgxNativeDateModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxNativeDateModule });
NgxNativeDateModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxNativeDateModule_Factory(t) { return new (t || NgxNativeDateModule)(); }, providers: [
        { provide: NgxMatDateAdapter, useClass: NgxMatNativeDateAdapter },
    ], imports: [[PlatformModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxNativeDateModule, { imports: function () { return [PlatformModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxNativeDateModule, [{
        type: NgModule,
        args: [{
                imports: [PlatformModule],
                providers: [
                    { provide: NgxMatDateAdapter, useClass: NgxMatNativeDateAdapter },
                ]
            }]
    }], function () { return []; }, null); })();
    return NgxNativeDateModule;
}());
export { NgxNativeDateModule };
var ɵ0 = NGX_MAT_NATIVE_DATE_FORMATS;
var NgxMatNativeDateModule = /** @class */ (function () {
    function NgxMatNativeDateModule() {
    }
NgxMatNativeDateModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxMatNativeDateModule });
NgxMatNativeDateModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxMatNativeDateModule_Factory(t) { return new (t || NgxMatNativeDateModule)(); }, providers: [{ provide: MAT_DATE_FORMATS, useValue: ɵ0 }], imports: [[NgxNativeDateModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxMatNativeDateModule, { imports: [NgxNativeDateModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxMatNativeDateModule, [{
        type: NgModule,
        args: [{
                imports: [NgxNativeDateModule],
                providers: [{ provide: MAT_DATE_FORMATS, useValue: ɵ0 }]
            }]
    }], function () { return []; }, null); })();
    return NgxMatNativeDateModule;
}());
export { NgxMatNativeDateModule };
export { ɵ0 };
