// @angular-material-components/moment-adapter/esm2015/lib/moment-adapter.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/moment-adapter.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgxMatDateAdapter, NGX_MAT_DATE_FORMATS } from '@angular-material-components/datetime-picker';
import { NgModule } from '@angular/core';
import { MAT_DATE_LOCALE } from '@angular/material/core';
import { NgxMatMomentAdapter, NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS } from './moment-adapter';
import { NGX_MAT_MOMENT_FORMATS } from './moment-formats';
import * as ɵngcc0 from '@angular/core';
export class NgxMomentDateModule {
}
NgxMomentDateModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxMomentDateModule });
NgxMomentDateModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxMomentDateModule_Factory(t) { return new (t || NgxMomentDateModule)(); }, providers: [
        {
            provide: NgxMatDateAdapter,
            useClass: NgxMatMomentAdapter,
            deps: [MAT_DATE_LOCALE, NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS]
        }
    ] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxMomentDateModule, [{
        type: NgModule,
        args: [{
                providers: [
                    {
                        provide: NgxMatDateAdapter,
                        useClass: NgxMatMomentAdapter,
                        deps: [MAT_DATE_LOCALE, NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS]
                    }
                ]
            }]
    }], null, null); })();
const ɵ0 = NGX_MAT_MOMENT_FORMATS;
export class NgxMatMomentModule {
}
NgxMatMomentModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxMatMomentModule });
NgxMatMomentModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxMatMomentModule_Factory(t) { return new (t || NgxMatMomentModule)(); }, providers: [{ provide: NGX_MAT_DATE_FORMATS, useValue: ɵ0 }], imports: [[NgxMomentDateModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxMatMomentModule, { imports: [NgxMomentDateModule] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxMatMomentModule, [{
        type: NgModule,
        args: [{
                imports: [NgxMomentDateModule],
                providers: [{ provide: NGX_MAT_DATE_FORMATS, useValue: ɵ0 }]
            }]
    }], null, null); })();
export { ɵ0 };
